pilot/pkg/xds/internalgen.go

```diff
@@ -142,14 +155,111 @@ func (sg *InternalGen) startPush(typeURL string, data []proto.Message) {
 // We can also expose ACKS.
 func (sg *InternalGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
 	res := []*any.Any{}
-	switch w.TypeUrl {
+
+	// Look for ?query parameters.  We don't use url.Parse() because we have no scheme
+	typeURL := w.TypeUrl
+	var qparams url.Values
+	if qIndex := strings.Index(w.TypeUrl, "?"); qIndex >= 0 {
+		typeURL = w.TypeUrl[:qIndex]
+		var err error
+		qparams, err = url.ParseQuery(w.TypeUrl[qIndex+1:])
+		if err != nil {
+			adsLog.Infof("Invalid TypeUrl query params: %q", w.TypeUrl)
+		}
+	}
+
+	switch typeURL {
 	case TypeURLConnections:
 		sg.Server.adsClientsMutex.RLock()
 		// Create a temp map to avoid locking the add/remove
 		for _, v := range sg.Server.adsClients {
 			res = append(res, util.MessageToAny(v.xdsNode))
 		}
 		sg.Server.adsClientsMutex.RUnlock()
+	case TypeDebugSyncronization:
+		res = sg.debugSyncz()
+	case TypeDebugConfigDump:
+		proxyID := qparams.Get("proxyID")
+		if proxyID == "" {
+			adsLog.Info("Config Dump w/o proxyID query parameter")
+			break
+		}
+		res = sg.debugConfigDump(proxyID)
+	default:
+		adsLog.Infof("Unknown TypeUrl: %q", w.TypeUrl)
+	}
+	return res
+}
+
+func (sg *InternalGen) debugSyncz() []*any.Any {
+	res := []*any.Any{}
+
+	sg.Server.adsClientsMutex.RLock()
+	for _, con := range sg.Server.adsClients {
+		con.mu.RLock()
+		// Skip "nodes" without metdata (they are probably istioctl queries!)
```

Comment:
We do not.  In a review of the `istioctl x version` command, @costinm expressed that it would be better if there was a way, but said this method would work.  So I have continued using it.

I am not up-to-speed with the logic for how Istiod tracks connections.  My thought is that if _istioctl_ and sidecars use the same XDS interface, and _istioctl_ doesn't shouldn't be tracked as a connection, then the XDS interface should provide a way for a connection to indicate this, and the logic should be removed from here.

The two commands I implemented, `version` and `proxy-status`, are all one-shots.  My expectation is that you, @therealmitchconnors , will be doing `wait` and this may have a long-running XDS connection to Istiod.  We might want to be able to see how many debug clients are in such long-running connections.

If all you are asking is for me to wrap that logic in a method `isReallyANode()` that is easy, anything else requires discussion.

Reply:
Fixed with `isSidecar(*Connection)` method.  I'll leave it to others to improve that method.